[READY] Always enable spell checking for better FixIts

puremourning opened the pull request:
Fixes #546 

We enable spell-checking to get better FixIt suggestions. The only time we don't is if the user's flags explicitly request it is disabled.

The change is in `PrepareFlagsForClang` because that is run both when getting the flags from the `FlagsForFile` function and when supplying them in the request. However, rather than having to update every test which does this to also expect `-fspell-checking` in the result, we use the same approach as `FlagsForFile` to allow the test system to disable adding it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/548)
<!-- Reviewable:end -->

micbou commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/completers/cpp/flags.py, line 409 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMptKhYDncJtyZ74fnh:-KMptKhZkzzujRiI_dqF:307176134) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/completers/cpp/flags.py#L409)):*
> ```Python
>   # a feature useful enough that we just always turn it on. That is, unless
>   # the user explicitly turned it off (with -fno-spell-checking).
>   if '-fno-spell-checking' in flags:
> ```

We could check for `-fspell-checking` too so that we don't add it a second time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/548#-:-KMpw1ub4LOUXm4xA_V8:1417088721)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/completers/cpp/flags.py, line 409 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMptKhYDncJtyZ74fnh:-KMpwmDemAndazgETo46:-345752202) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/completers/cpp/flags.py#L409)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We could check for `-fspell-checking` too so that we don't add it a second time.
</details>
true, i suppose, but i think adding it 2x is harmless.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/548)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Thanks!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 406 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMqUGU__PAfjt0mzMCV:-KMqUGU__PAfjt0mzMCW:1535222447) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/completers/cpp/flags.py#L406)):*
> ```Python
>   # hints (in the form of fix-its) in the case of certain diagnostics. A common
>   # example is "no type named 'strng' in namespace 'std'; Did you mean
>   # 'string'? (FixIt)". This is enabled by default in the clang driver and is
> ```

It might be a bit confusing that you say that it's enabled by default and yet we add it ourselves. The difference is between the clang driver and calling libclang, but that might not be obvious to the reader.

I'd suggest clarifying this a bit to avoid confusion.

---

*[ycmd/completers/cpp/flags.py, line 409 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMptKhYDncJtyZ74fnh:-KMqUVAqOiRLY-ml073H:328195815) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/completers/cpp/flags.py#L409)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary>
> true, i suppose, but i think adding it 2x is harmless.
</details>
Personally, I don't care either way since there's zero chance of the duplicate causing problems.

---

*[ycmd/tests/clang/testdata/FixIt_Clang_cpp11.cpp, line 68 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMqUkw036gkEVJbIDdM:-KMqUkw036gkEVJbIDdN:5173007) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/tests/clang/testdata/FixIt_Clang_cpp11.cpp#L68)):*
> ```cpp
> 
> namespace Typo {
>     struct SpellingIsNotMyStrongPoint;
> ```

:D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/548#-:-KMqUnyLJyBndVUAvN2J:-244862327)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/548#-:-KMsUGaHgcB8eyUNxKpZ:980077423)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #547) made this pull request unmergeable. Please resolve the merge conflicts.

puremourning commented:
Rebased. I think this is ready to go.

---

Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 406 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/548#-KMqUGU__PAfjt0mzMCV:-KMt0NghWIKyp0y8SS9V:-219674365) ([raw file](https://github.com/valloric/ycmd/blob/f7a31e9475f9c9056c14345164f09c431ef51e88/ycmd/completers/cpp/flags.py#L406)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> It might be a bit confusing that you say that it's enabled by default and yet we add it ourselves. The difference is between the clang driver and calling libclang, but that might not be obvious to the reader.
> 
> I'd suggest clarifying this a bit to avoid confusion.
</details>
I thought 'clang driver' made that clear, but i have added a further clarification:

```
  # "Typo correction" (aka spell checking) in clang allows it to produce
  # hints (in the form of fix-its) in the case of certain diagnostics. A common
  # example is "no type named 'strng' in namespace 'std'; Did you mean
  # 'string'? (FixIt)". This is enabled by default in the clang driver (i.e. the
  # 'clang' binary), but is not when using libclang (as we do). It's a useful
  # enough feature that we just always turn it on unless the user explicitly
  # turned it off in their flags (with -fno-spell-checking).
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/548#-:-KMuOhfsjgiBwQEzIB5C:1527044772)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Awesomesauce!

@homu r+

homu commented:
:pushpin: Commit 48c2612 has been approved by `Valloric`

<!-- @homu r=Valloric 48c26125affcae8112e8e89b5a2e0541a3384671 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/145398105)
